internal/pkg/stream/ecs.go

```diff
@@ -0,0 +1,161 @@
+}
+
+// ECSDeploymentStreamer is a Streamer for ECSService descriptions until the deployment is completed.
+type ECSDeploymentStreamer struct {
+	client                 ECSServiceDescriber
+	cluster                string
+	service                string
+	deploymentCreationTime time.Time
+
+	subscribers   []chan ECSService
+	done          chan struct{}
+	pastEventIDs  map[string]bool
+	eventsToFlush []ECSService
+}
+
+// NewECSDeploymentStreamer creates a new ECSDeploymentStreamer that streams service descriptions
+// since the deployment creation time and until the primary deployment is completed.
+func NewECSDeploymentStreamer(ecs ECSServiceDescriber, cluster, service string, deploymentCreationTime time.Time) *ECSDeploymentStreamer {
+	return &ECSDeploymentStreamer{
+		client:                 ecs,
+		cluster:                cluster,
+		service:                service,
+		deploymentCreationTime: deploymentCreationTime,
+		done:                   make(chan struct{}),
+		pastEventIDs:           make(map[string]bool),
+	}
+}
+
+// Subscribe returns a read-only channel that will receive service descriptions from the ECSDeploymentStreamer.
+func (s *ECSDeploymentStreamer) Subscribe() <-chan ECSService {
+	c := make(chan ECSService)
+	s.subscribers = append(s.subscribers, c)
+	return c
+}
+
+// Fetch retrieves and stores ECSService descriptions since the deployment's creation time
+// until the primary deployment's running count is equal to its desired count.
+// If an error occurs from describe service, returns a wrapped err.
+// Otherwise, returns the time the next Fetch should be attempted.
+func (s *ECSDeploymentStreamer) Fetch() (next time.Time, err error) {
+	out, err := s.client.Service(s.cluster, s.service)
+	if err != nil {
+		return next, fmt.Errorf("fetch service description: %w", err)
+	}
+	var deployments []ECSDeployment
+	for _, deployment := range out.Deployments {
+		status := aws.StringValue(deployment.Status)
+		desiredCount, runningCount := aws.Int64Value(deployment.DesiredCount), aws.Int64Value(deployment.RunningCount)
+		deployments = append(deployments, ECSDeployment{
+			Status:          status,
+			TaskDefRevision: parseRevisionFromTaskDefARN(aws.StringValue(deployment.TaskDefinition)),
+			DesiredCount:    int(desiredCount),
+			RunningCount:    int(runningCount),
+			FailedCount:     int(aws.Int64Value(deployment.FailedTasks)),
+			PendingCount:    int(aws.Int64Value(deployment.PendingCount)),
+			RolloutState:    aws.StringValue(deployment.RolloutState),
+		})
+		if status == ecsPrimaryDeploymentStatus && desiredCount == runningCount {
+			// The deployment is done, notify that there is no need for another Fetch call beyond this point.
+			close(s.done)
+		}
+	}
+	var failureMsgs []string
+	for _, event := range out.Events {
+		if createdAt := aws.TimeValue(event.CreatedAt); createdAt.Before(s.deploymentCreationTime) {
+			break
+		}
+		id := aws.StringValue(event.Id)
+		if _, ok := s.pastEventIDs[id]; ok {
+			break
+		}
```

Comment:
Why do we need this validation? Isn't creation time validation enough?

Reply:
ooh good question, it just depends on what you want the stream to emit. 
Currently, we only emit new failure messages in between events. If we don't store `pastEventIDs`, we would write all failure messages for the deployment for each event. Which has the nice pro of less code, but then you don't shortcircuit if you see the same event twice. I thought this was closer to what a client of a `ECSDeploymentStreamer` might expect.

Maybe a simpler thing to do would be just storing `lastSeenEventID` instead of a map since the events are ordered